### ISSUE TITLE
Add connect, read, and write timeouts to client

### ISSUE THREAD

**Issue by sfackler** [hyperium/hyper#1234]

**Comment by hjr3** [comment]
I am going to try get a PR up for this. There are two options for timers, but I am unsure of which is best for hyper. The [tokio_timer](https://github.com/tokio-rs/tokio-timer) crate is more accurate but uses a dedicated thread. The [Timeout](https://tokio-rs.github.io/tokio-core/tokio_core/reactor/struct.Timeout.html) in tokio-core is less accurate but does not require the extra thread.

If I was making the decision, I would create a single tokio-timer Timer (1 dedicated thread), configure it properly and then use that for any timeout needs. Does anyone else have any opinions?

**Comment by seanmonstar** [comment]
hyper 0.11 didn't ship with these for a couple reasons:

- It's possible to have a rough overall timeout by using timeouts outside of hyper.
- I didn't want to pick a timer implementation that was different than the user, and then end up that multiple timer implementations were running.

Still, I realize it would be useful to have these in hyper, so let's explore!

- For anyone wanting it immediately, you can largely get most of it working externally. A "connect" timeout can be used by providing a custom `Connect` that wraps the `HttpConnector` and a `Timeout`. A "write" timeout (for a body) can be sort of hacked in by providing a custom `Body` stream. And a "read" timeout could be used by selecting on the `FutureResponse`, and/or the response body. I realize it's not super accurate, but it's there for the urgent.
- For the implementation in hyper, unless there the coarser timeouts are bad (it shouldn't be seconds off, just perhaps a few milliseconds if the timer isn't quite reached, and another loop of the `Core` would take a few milliseconds to get back), I'm leaning towards using `tokio-core` timeouts. They're lighter weight, doesn't start a new thread, and doesn't require another dependency.

**Comment by sfackler** [comment]
I wrote up a little crate handling read and write timeouts via the standard tokio-core timer: https://crates.io/crates/tokio-io-timeout

**Comment by nielsle** [comment]
Withoutboats has a pull request that adds middleware wrappers to 'tokio_service::Service'. Perhaps that could be used to add logging and timing and retries to the 'hyper::Client'

tokio-rs/tokio-service#21

This solution would be fairly generic and extendable, but it might also be slightly complicated.

**Comment by hjr3** [comment]
Based on the discussion, I wrote a `TimeoutConnector` that wraps an existing `Connect` implementation here https://github.com/hyperium/hyper/compare/master...hjr3:issue-1234?expand=1. I left the code in the `connect.rs` file for now, so as not to deal with module shenanigans. The `TimeoutConnector` can be used with the `HttpsConnector` in [hyper_tls](https://github.com/hyperium/hyper-tls) too. This may also get use closer to the suggestion from @nielsle about using Middleware. I say _may_ because I did not like very closely that how the PR works. I still owe some tests.

The alternative is to embed the timeout in `HttpConnector` too, but the wrapper seemed better/cleaner.

I will start playing around with the crate from @sfackler and see if I can get read/write timeouts setup.

**Comment by skinowski** [comment]
Could you please update the 'advanced' client example with an overall timeout? This is the general case and if it's in the example, it'll be super helpful.

**Comment by hjr3** [comment]
@skinowski Definitely. Where is the _advanced_ example? I only see the [simple](https://github.com/hyperium/hyper/blob/master/examples/client.rs) example.

**Comment by skinowski** [comment]
@hjr3 Here: https://hyper.rs/guides/client/advanced/

I've also figured how to do this myself:

let dur = Duration::from_millis(1000);
let work = client.get(...).and_then(...)....;
let tm = timer.timeout(work, dur);

But it's really easy to get caught in wrapped types of these futures: map, and_then, map_err, select2, join, etc. are not easy to follow for a beginner.

**Comment by hjr3** [comment]
I have read/write timeouts in place. After putting those in, it seems natural to want to configure the client with a `connect_timeout` instead of wrapping the `Connect` implementation. I am trying to figure out how to do that.

**Comment by hjr3** [comment]
@skinowski I also added a PR to add the general timeout to the guide https://github.com/hyperium/hyperium.github.io/pull/13

**Comment by hjr3** [comment]
I created https://crates.io/crates/hyper-timeout to get connect, read and write timeout functionality. So far, it has been working for my use case. There may be some changes required to tokio-io-timeout, including this comment from @seanmonstar:

> As a side note, looking at the tokio-io-timeout crate, the implementations of AsyncRead and AsyncWrite don't pass forward the read_buf and write_buf methods, which would mean readv and writev are no longer used for TcpStream.

I will see about getting some PRs up for tokio-io-timeout.

**Comment by sfackler** [comment]
I've added implementations of `read_buf` and `write_buf`.

**Comment by hjr3** [comment]
@sfackler 👍 i will update my crate.

**Comment by softprops** [comment]
I was just going to open an issue about server side read write timeouts. Would any of the potential solutions above also cover servers?

**Comment by sfackler** [comment]
@softprops the tokio-io-timeout crate can handle them if you manage the server accept loop manually. Example: https://github.com/sfackler/foo/blob/master/src/main.rs#L53

**Comment by softprops** [comment]
Nice. What are the changes of something like this landing in hyper itself?

**Comment by seanmonstar** [comment]
The current situation of timers in tokio are [in flux](https://github.com/tokio-rs/tokio-rfcs/pull/2), so I have been wanting to wait for the dust there to settle.

**Comment by softprops** [comment]
Makes sense

**Comment by cetra3** [comment]
Is there still plans to add read/write timeouts to client requests?  I understand you can use tokio's timeout function to timeout the request overall, but this is sometimes not what you want.

Sometimes a download can be slow as heck, but it's still downloading, so as long as the time between delivery of bytes is not hit, then it's an "active" connection.

**Comment by hjr3** [comment]
@cetra3 Hi, you can look at using https://crates.io/crates/hyper-timeout. I just updated it to work with hyper v0.12.

**Comment by cetra3** [comment]
Cheers @hjr3, I'm assuming this is compatible with the https connector?

**Comment by hjr3** [comment]
Yes, the example client uses the https connector.

**Comment by chewi** [comment]
We've been using hyper-timeout for a little while but it has one major drawback that only an implementation within hyper can provide. If the requested host name resolves to multiple IP addresses, hyper internally tries to connect to each one in turn. If one of them doesn't respond at all then it hangs until the OS-level timeout kicks in. This defaults to 127 seconds on Linux due to `tcp_syn_retries` being 6. hyper-timeout wraps around all the connection attempts rather than each one individually so setting that timeout to something shorter means that it will give up before trying the second address. You could reduce `tcp_syn_retries` but this is kernel-wide so this is far from ideal.

**Comment by let4be** [comment]
Is there any update on this?... 
fine grained control of timeout(including connecting to several IPs as mentioned by @chewi) would be really nice to have

**Comment by chewi** [comment]
Yes, Hyper supports this natively now. It was added in Hyper 13 and I backported it to 12.

**Comment by EdorianDark** [comment]
It it is supported, why is this issue not closed?
